### ISSUE TITLE
[pom] Maven POM Rules don't check files named "pom.xml"

### ISSUE THREAD

**Issue by kris-scheibe** [pmd/pmd#1540]
**Affects PMD Version:** 
6.10.0

**Rule:**
all rules for "Maven POM", tested with [ProjectVersionAsDependencyVersion](https://pmd.github.io/latest/pmd_rules_pom_errorprone.html#projectversionasdependencyversion)

**Description:**
When running PMD on Maven POM files (by convention named `pom.xml`), the files aren't being picked up for validation.

The `PomLanguageModule` seems to define Maven POM files as XML files having a "pom" extension.
Is there a configuration that makes PMD pick up a `pom.xml` instead of looking for a `something.pom`?

**Code Sample demonstrating the issue:**

Given this example `pom.xml` (a failing example taken from the PMD unit tests)
```xml
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>net.sourceforge.pmd</groupId>
    <artifactId>xml-pom</artifactId>
    <version>1.0.0-SNAPSHOT</version>

    <dependencies>
        <dependency>
            <groupId>x.y.z</groupId>
            <artifactId>z</artifactId>
            <version>${project.version}</version>
        </dependency>
    </dependencies>
</project>
```
and running PMD with
```bash
# should report the dependency version in ~/pmd-pom-test/pom.xml
pmd-bin-6.10.0/bin$ ./run.sh pmd -no-cache -d ~/pmd-pom-test/ -R category/pom/errorprone.xml/ProjectVersionAsDependencyVersion
```
Doesn't show any violations.

Renaming the file to `xml.pom` (or really anything ending with `.pom`) does report the violation:
```bash
# does report the dependency version in ~/pmd-pom-test/xml.pom
pmd-bin-6.10.0/bin$ ./run.sh pmd -no-cache -d ~/pmd-pom-test/ -R category/pom/errorprone.xml/ProjectVersionAsDependencyVersion
/home/kris/pmd-pom-test/xml.pom:11:     Do not use project.version to express a dependency version.
```

**Running PMD through:** *CLI*

**Comment by kris-scheibe** [comment]
OK, so I've dug a little deeper and it seems that PMD completely relies on the extension for the file-to-language mapping.
Checking further: Without breaking compatibility (i.e. changing the definition of `net.sourceforge.pmd.lang.Language`) making the `PomLanguageModule` pick up `pom.xml` files doesn't work without some ugly hacks in the language discovery mechanism in pmd-core, essentially returning the extension of `pom.xml` as `pom` and when filtering for `*.pom` files also allowing `pom.xml`...

I made the checks work in my builds by automatically copying `pom.xml` to `pmd.pom` and replacing the file name in the generated report.

Reworking the language discovery mechanism might be a candidate for **PMD 7.x**, though:
- Add a `getFilePatterns()` to the `Language` interface to return glob-style patterns of supported files. With Java 8 this can be non-breaking change by adding a default implementation to return an empty list.
- For backwards-combatibility: interpret the existing extensions as `*.<extension>` in addition to defined file patterns.
- Change the language discovery and file filtering to use the file patterns instead of extensions.

**Comment by oowekyala** [comment]
@kris-scheibe Your analysis is correct, and your proposed change would be welcome in PMD 7.0. Btw, it's been a long standing feature request to make the extension/file pattern mapping configurable by the user (see #461), so I think we should probably design a solution with both issues in mind. For the moment though, I think #461 has too many open questions to start implementing on any of those two fronts (see my recent comment on it)

**Comment by rsoesemann** [comment]
@oowekyala it just appeared that the Salesforce community could make enormous use of the XML engine and its XPath rules if this would be implemented. Why? Because besides Apex most of the stuff you do in Salesforce is so called XML metadata with well defined nodes.

Reading through all the history you mentioned I thought why not add this pragmatic extension to PMD 7 with now compatibility issues: Add an extensions attribute to custom rules like this

```xml
    <rule name="CustomFieldNamingConvention" 
        language="xml" 
        extensions=".flow,.meta-xml,.object"
        message="Incorrectly named field" 
        class="net.sourceforge.pmd.lang.rule.XPathRule">
        <properties>
            <property name="xpath">
                <value><![CDATA[
//CustomField[
                ]]></value>
            </property>
        </properties>
    </rule>
```
if the extensions attribute is not defined everything works as before. If defined rules are only applied by the Tree Walker if the extension matches.
